### PR TITLE
feat: Support deployment and verification of upgradeable SpokePools

### PR DESCRIPTION
**Pull request by nicholaspai**
`hardhat-upgrades` has some useful libraries we can use to deploy and verify proxies. I've tested these all out on prod.

### CODE REVIEW

File: deploy/003_deploy_optimism_spokepool.ts [hunk 1]
```diff
@@ -1,26 +1,33 @@
 import "hardhat-deploy";
-import { HardhatRuntimeEnvironment } from "hardhat/types/runtime";
+import hre from "hardhat";
+import { getContractFactory } from "../test/utils";
```
Comment (pxrl) [inline deploy/003_deploy_optimism_spokepool.ts:3]: Given that we're going to rely on this for prod, any thoughts on moving it from `src/test/utils` => `src/utils` ?

File: deploy/005_deploy_arbitrum_spokepool.ts [hunk 2]
```diff
@@ -16,17 +13,30 @@ const func = async function (hre: HardhatRuntimeEnvironment) {
...
+  console.log(`SpokePool deployed @ ${instance.address}`);
+  const implementationAddress = await upgrades.erc1967.getImplementationAddress(instance.address);
+  console.log(`Implementation deployed @ ${implementationAddress}`);
+
+  // hardhat-upgrades overrides the `verify` task that ships with `hardhat` so that if the address passed
+  // is a proxy, hardhat will first verify the implementation and then the proxy and also link the proxy
+  // to the implementation's ABI on etherscan.
+  // https://docs.openzeppelin.com/upgrades-plugins/1.x/api-hardhat-upgrades#verify
+  await run("verify:verify", {
+    address: instance.address,
```
Comment (pxrl) [inline deploy/005_deploy_arbitrum_spokepool.ts:39]: This part seems pretty repetitive across the various deployment scripts - might be ripe for a utility function that handles all this. wdyt?

File: test/SpokePool.SlowRelay.ts [hunk 3]
```diff
@@ -456,7 +456,7 @@ describe("SpokePool Slow Relay Logic", async function () {
               slowFills.find(
                 (slowFill) =>
                   slowFill.relayData.destinationToken === destErc20.address &&
-                  slowFill.payoutAdjustmentPct === ethers.utils.parseEther("101").toString()
+                  slowFill.payoutAdjustmentPct.toString() === ethers.utils.parseEther("101").toString()
```
Comment (pxrl) [inline test/SpokePool.SlowRelay.ts:459]: Any reason to not use the native BN comparison?

```suggestion
                  slowFill.payoutAdjustmentPct.eq(ethers.utils.parseEther("101"))
```

File: deploy/003_deploy_optimism_spokepool.ts (hunk 1 above)
Comment (nicholaspai) [reply deploy/003_deploy_optimism_spokepool.ts:3]: Great call these all seem like useful functions to have in `src`

File: deploy/005_deploy_arbitrum_spokepool.ts (hunk 2 above)
Comment (nicholaspai) [reply deploy/005_deploy_arbitrum_spokepool.ts:39]: yeah thats reasonable

**Review by mrice32 — APPROVED** [review]

File: test/SpokePool.SlowRelay.ts [hunk 4]
```diff
@@ -485,7 +485,7 @@ describe("SpokePool Slow Relay Logic", async function () {
               slowFills.find(
                 (slowFill) =>
                   slowFill.relayData.destinationToken === weth.address &&
-                  slowFill.payoutAdjustmentPct === ethers.utils.parseEther("-1.01").toString()
+                  slowFill.payoutAdjustmentPct.toString() === ethers.utils.parseEther("-1.01").toString()
```
Comment (pxrl) [inline test/SpokePool.SlowRelay.ts:488]: Spotted one more:

```suggestion
                  slowFill.payoutAdjustmentPct.eq(ethers.utils.parseEther("-1.01"))
```

**Review by pxrl — APPROVED** [review]
LGTM

File: deployments/deployments.json [hunk 5]
```diff
@@ -7,7 +7,7 @@
     "Arbitrum_SendTokensAdapter": { "address": "0xC06A68DF12376271817FcEBfb45Be996B0e1593E", "blockNumber": 16691987 },
     "Boba_Adapter": { "address": "0x33B0Ec794c15D6Cc705818E70d4CaCe7bCfB5Af3", "blockNumber": 14716798 },
     "Ethereum_Adapter": { "address": "0x527E872a5c3f0C7c24Fe33F2593cFB890a285084", "blockNumber": 14704381 },
-    "SpokePool": { "address": "0x4D9079Bb4165aeb4084c526a32695dCfd2F77381", "blockNumber": 14819486 },
+    "SpokePool": { "address": "0x7CE9f4189d0b0b04B834a5BDeDE8E518Ad1d7dC0", "blockNumber": 17069252 },
```
Comment (james-a-morris) [inline deployments/deployments.json:10]: QQ: will we need to be mindful of the data from the old spoke pool after a sufficiently large number of periods?

File: contracts/SpokePool.sol [hunk 6]
```diff
@@ -1223,24 +1223,28 @@ abstract contract SpokePool is
...
                     msg.sender,
-                    relayData.recipient,
+                    relayExecution.updatedRecipient,
+                    amountToSend
+                );
+            else
+                IERC20Upgradeable(relayData.destinationToken).safeTransfer(
+                    relayExecution.updatedRecipient,
                     amountToSend
                 );
```
Comment (james-a-morris) [inline contracts/SpokePool.sol:1240]: Should we modify the `tsconfig.json` such that we enforce brackets after control statements?

**Review by james-a-morris — COMMENTED** [review]
Looks good, left a few comments 🙂

File: deployments/deployments.json (hunk 5 above)
Comment (nicholaspai) [reply deployments/deployments.json:10]: We will not have to care about that data anymore, what do you mean?

File: contracts/SpokePool.sol (hunk 6 above)
Comment (nicholaspai) [reply contracts/SpokePool.sol:1240]: Yes, can you recommend how we do that?

File: contracts/SpokePool.sol [hunk 7]
```diff
@@ -1193,6 +1193,13 @@ abstract contract SpokePool is
             "invalid repayment chain"
         );
 
+        // If this is a partial fill, require that the fill has no message. This prevents situations in which
+        // a message is sent multiple times for multiple partial fills.
+        require(
+            relayExecution.relay.amount == fillAmountPreFees || relayExecution.updatedMessage.length == 0,
+            "invalid partial fill message"
+        );
+
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply contracts/SpokePool.sol:1202]: yeah I will move this to a separate PR i only added it here temporarily

File: package.json [hunk 8]
```diff
@@ -43,6 +43,7 @@
     "arb-bridge-peripherals": "^1.0.5"
   },
   "devDependencies": {
+    "@across-protocol/sdk-v2": "0.5.0-beta.16",
```
Comment (pxrl) [inline package.json:46]: Hm, why are we importing the SDK here? The SDK also depends on contracts-v2 and it looks like it's creating circular dependencies in `yarn.lock`.

Comment (nicholaspai) [reply package.json]: Removed

**Pull request merged**